### PR TITLE
ADD subscription expression evaluation

### PR DESCRIPTION
**Pull request by fgalan**
Implements part of #1316

@kzangeli 
@crbrox

### CODE REVIEW

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 1]
```diff
@@ -909,14 +1304,29 @@ static bool processAreaScope(const Scope* scoP, BSONObj &areaQuery)
...
+*   be applied on MongoDB. The return value is not used. This is the typical use for queryContext and derived
+*   operations
+* - if cerP is not NULL, then the filters are evaluated in memory on the cerP. The return value is used to
+*   specify if cerP match the filter or not. The filters parameter is not used (although the caller needs to
+*   provide it in order to conform with function signature).
+*
+* I don't like too much to have a function with two quite parallel behaviours in the same logic. Probably a
+* smarter design will be to divide this function in two part: the one focused in string parsing and the one
+* focused in appliying filters to each different sytem (either in memory or as BSON).
+*
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Opinions on the above comment and ideas to improve the current implementation are welcome :)

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_combined.test [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+echo "02. Create entity with temperature=21 status=on (not notif)"
+echo "==========================================================="
+payload='{
+  "id": "E1",
+  "type": "T",
+  "temperature": 21,
+  "status": "on"
+}'
+orionCurl --url /v2/entities --payload "$payload" --json
```
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_combined.test]: Semi-offtopic of this PR but... I wonder how (now that #1259 has landed in develop) this can work. In order to work, it should include `?options=keyValues`.

Maybe something is missing in #1259?

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_enum_string.test [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+#
+# 01. Create subscription with q: colour==black,white
+# 02. Create entity with colour=green (not notif)
+# 03. Update entity to colour=white (notif)
+# 04. Update entity to colour=red (not notif)
+# 05. Update entity to colour=black (notif)
+# 06. Dump accumulator (2 notifications)
+#
+
+echo "01. Create subscription with q: colour!=black,white"
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_enum_string.test]: !=  => ==

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_single_number.test [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+#
+# 01. Create subscription with q: status==22
+# 02. Create entity with status=23 (not notif)
+# 03. Update entity to status=22 (notif)
+# 04. Update entity to status=24 (not notif)
+# 05. Update entity to status=22 (notif)
+# 06. Dump accumulator (2 notifications)
+#
+
+echo "01. Create subscription with q: status==on"
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_single_number.test]: on => 22

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_or_equal_than.test [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+POST Subscriptions V2 q expression (lesser or equal than)
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_or_equal_than.test]: lesser => less

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_than.test [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+POST Subscriptions V2 q expression (lesser than)
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_than.test]: lesser => less

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_nocache.test [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+--SHELL--
+
+#
+# 01. Create subscription with q: temperature==20..40
+# 02. Create entity with t=10 (not notif)
+# 03. Update entity to t=15 (not notif)
+# 04. Update entity to t=25 (notif)
+# 05. Update entity to t=35 (notif)
+# 06. Update entity to t=45 (not notif)
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_nocache.test]: Might be interesting to test corner cases 20 and 40 here.
If something ever goes wrong here, the probability of failure is always higher in the corner cases

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_number.test [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+echo "05. Update entity to avg=93 (notif)"
+echo "==================================="
+payload='{
+  "avg": 93
+}'
+orionCurl --url /v2/entities/E1 --payload "$payload" --json -X PATCH
+echo
+echo
+
+
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_number.test]: You only use integers in this test.
Yet, the test is called 'not enum string' ...
It should work with strings as well, right?

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Well, you have already split the function in two. Very little common code, just go all the way (which means we'll have some code duplicated). I don't see any problem with that.
I also don't see any problem in keeping it the way it is now. A 'fork function'.
What is not very nice is the 'two out of three parameters', that's all.

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_string.test [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+brokerStart CB 0 IPv4 -noCache
+accumulatorStart
+
+--SHELL--
+
+#
+# 01. Create subscription with q: colour!=black,white
+# 02. Create entity with colour=black (not notif)
+# 03. Update entity to colour=red (notif)
+# 04. Update entity to colour=while (not notif)
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_string.test]: white? :-)

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_single_number.test [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+--SHELL-INIT--
+dbInit CB
+# FIXME P10: replace with just 'brokerStart CB' once cache part gets implemented
+brokerStart CB 0 IPv4 -noCache
+accumulatorStart
+
+--SHELL--
+
+#
+# 01. Create subscription with q: status!=22
```
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_single_number.test]: This should notify with either number or string, right?

Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_single_number.test]: A test with status!='22'  would be nice.
I am just not sure that has been implemented yet

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 11]
```diff
@@ -1553,10 +1553,8 @@ static bool matchExpression(ContextElementResponse* cerP, const std::string& q)
     return true;
   }
 
-  //TBD (issue #1316). We would need a variant of the qStringFilters function, but
-  // appliying the filter to cerP in memory instead of generating a BSON to be used
-  // with mongo
-  return true;
+  std::vector<BSONObj> filters;   // not really used, just to match qStringFilters signature
+  return qStringFilters(q, filters, cerP);
```
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: This here is the only part I don't like about the current implementation.

Here *qStringFilters* is called like this:

     qStringFilters(q, NOTHING, cerP);

and later on it is called like this:

    qStringFilters(q, filters, NOTHING);

It's either this or a few lines copied.

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_combined.test (hunk 2 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_combined.test]: Moved to issue https://github.com/telefonicaid/fiware-orion/issues/1688

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_or_equal_than.test (hunk 5 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_or_equal_than.test]: I'm using the same wording that the one in http://telefonicaid.github.io/fiware-orion/api/v2/ for the Simple Query Language section.

If this is wrong, it should be changed also in that document (a bit surprising anyway, given that the .apib has been reviewed many times by many people and such typo should has been quickly spotted in that process...)

File: src/lib/ngsi/ContextElement.cpp [hunk 12]
```diff
@@ -112,6 +112,25 @@ std::string ContextElement::render(ConnectionInfo* ciP, RequestType requestType,
 
 /* ****************************************************************************
 *
+* ContextElement::getAttribute
+*/
+ContextAttribute* ContextElement::getAttribute(std::string attrName)
```
Comment (kzangeli) [inline src/lib/ngsi/ContextElement.cpp]: Would probably have been better to implement a 'lookup' method for ContextAttributeVector, but, never mind.

NTC

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_number.test (hunk 8 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_number.test]: I think the problem is that the name of the test is wrong (other test has the same problem). Fixing soon...

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_single_number.test (hunk 10 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_single_number.test]: This notifies only for number. As you suggest, in order to be based in string it should use status!='22'. You are right: this is not implemented (as far as I understand looking to qStringFilters function) and is the root cause of bugs likes this one https://github.com/telefonicaid/fiware-orion/issues/1595

However, ' should work to avoid vectors, e.g:

```
colour=='black,white','red,blue'
```

I'd include some tests for that.

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 11 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: I don't either like it :) However, any suggestion to improve?

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_or_equal_than.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_or_equal_than.test]: I have actually noted (and complained about) lesser before. Can't remember where exactly.
Let's look up lesser just to make sure I am not wrong. 
English is after all my third language ...

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_enum_string.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_enum_string.test]: Fixed in 39c2584

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_single_number.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_equal_single_number.test]: Fixed in 39c2584

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_than.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_lesser_than.test]: Fixed in 39c2584

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_nocache.test (hunk 7 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_nocache.test]: Fixed in 39c2584

Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_nocache.test]: (Also for the test with temperature!=20..40)

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_number.test (hunk 8 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_number.test]: Fixed in 39c2584

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_string.test (hunk 9 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_enum_string.test]: Fixed in 39c2584

File: test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_single_number.test (hunk 10 above)
Comment (fgalan) [inline test/functionalTest/cases/1316_create_subscription/post_subs_expr_q_not_equal_single_number.test]: Fixed in 39c2584

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 11 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Let's keep it the way it is

NTC

**Comment by crbrox** [comment]
LGTM

**Comment by kzangeli** [comment]
LGTM
